2023/09/testing-user-defined-functions-with-pgtap.md

```diff
@@ -0,0 +1,462 @@
+:description
+    A short description of the test.
+```
+
+`has_function()` is useful in checking the existence of a specific function. If you want to parse an entire schema, `functions_are()` is your command with pretty straightforward parameters.
+
+#### `functions_are()`
+Variations: 
+```postgresql
+SELECT functions_are( :schema, :functions, :description );
+SELECT functions_are( :schema, :functions );
+SELECT functions_are( :functions, :description );
+SELECT functions_are( :functions );
+```
+Example:
+```postgresql
+SELECT functions_are('your_schema', ARRAY['add_numbers', <All other functions need to be listed as well>]);   -- default description used
+```
+Parameters
+```
+:schema
+    Name of a schema in which to find functions.
+:functions
+    An array of function names.
+:description
+    A short description of the test.
+```
+#### `ok()` and `is()`
+
+`ok()` and `is()` have similar a function with slightly different parameters:
+```postgresql
+SELECT ok( :boolean, :description );
+```
+Example:
+```postgresql
+SELECT ok(add_numbers(5,5)=10, 'add_numbers correctly works');
+```
+Parameters
+```
+:boolean
+    A boolean value indicating success or failure.
+:description
+    A short description of the test.
+```
+```postgresql
+SELECT is(   :have, :want, :description );
+```
+Example:
+```postgresql
+SELECT is(add_numbers(5,5), 10, 'add_numbers correctly works');
+```
+
+You can find functions that cover pretty much anything in PostgreSQL here on [pgTAP's documentation site](https://pgtap.org/documentation.html) 
+
+
+### Testing more complex user-defined stored functions:
+
+An existing client product of ours, utilizes a specific function that also calls several other functions depending on its given argument. This function is called by the application
+and is passed an INTEGER as an argument from which the function bases an initial SELECT query’s WHERE condition on, in this case `WHERE id = [argument]’. Variables are then assigned
+column values from this queried table. These variables are then used as the conditions to `PERFORM` functions that change parts of the database.
+
+When testing this sort of scenario, it became apparent that I needed a set of test data, an ability to track that test data, run the function with the data, and test to see if the 
+changes are correctly, or incorrectly, performed. 
+
+The nice thing about pgTAP is that unit tests are meant to be rolled back leaving the database as you found it. When faced with a more complex function that nests other functions which
+loop through multiple tables, it may make more sense to fill those tables temporarily with the data required for the tests. 
+
+Let's take a nostalgia trip and imagine we’re operating a DVD rental store. The Pagila sample database, which you can download on GitHub, represents our scenario. It comes populated
+and we’ll construct a function which utilizes other functions which perform changes on multiple tables.
```

Comment:
"which utilizes other functions which perform changes on multiple tables" -> "which utilizes other functions that perform changes on multiple tables"